tools/gfx/vulkan/render-vk.cpp

```diff
@@ -2482,22 +2482,29 @@ class VKDevice : public RendererBase
         {
             auto& api = buffer->m_renderer->m_api;
             IBufferResource* stagingBuffer = nullptr;
-            transientHeap->allocateStagingBuffer(size, stagingBuffer, ResourceState::CopySource);
+            size_t stagingBufferOffset = 0;
+            transientHeap->allocateStagingBuffer(
+                size, stagingBuffer, stagingBufferOffset, MemoryType::Upload);
 
             BufferResourceImpl* stagingBufferImpl =
                 static_cast<BufferResourceImpl*>(stagingBuffer);
 
             void* mappedData = nullptr;
             SLANG_VK_CHECK(api.vkMapMemory(
-                api.m_device, stagingBufferImpl->m_buffer.m_memory, 0, size, 0, &mappedData));
-            memcpy(mappedData, data, size);
+                api.m_device,
+                stagingBufferImpl->m_buffer.m_memory,
+                0,
+                stagingBufferOffset + size,
+                0,
+                &mappedData));
+            memcpy((char*)mappedData + stagingBufferOffset, data, size);
             api.vkUnmapMemory(api.m_device, stagingBufferImpl->m_buffer.m_memory);
 
             // Copy from staging buffer to real buffer
             VkBufferCopy copyInfo = {};
             copyInfo.size = size;
             copyInfo.dstOffset = offset;
-            copyInfo.srcOffset = 0;
+            copyInfo.srcOffset = stagingBufferOffset;
```

Comment:
It's a minor thing, but looking at all these changes, it seems clear that we really need a type (probably even in the public header for `gfx`) to represent a pair of a buffer and offset:

```
// Needs thought on naming
struct BufferAndOffset
{
    IBufferResource* buffer;
    Offset offset;
};

// Probably needed until we can join the resource and view hierarchies
struct BufferViewAndOffset
{
    IBufferResourceView* view;
    Offset offset;
};

// Needed?
struct BufferSpan
{
    IBufferResource* buffer;
    Offset begin;
    Offset end; // or offset+size?
};
```

A lot of our public API functions that currently take a buffer+offset as distinct parameters should probably take one of those types, and then have convenience overloads that take just a buffer, or a buffer and offset pair. For C++ we can obviously make the `BufferAndOffset` type include an implicit-conversion constructor that takes a buffer pointer, which could eliminate the need for some of the wrappers.